azure/services/microsoft.dbformysql/resource.ftl

```diff
@@ -0,0 +1,319 @@
+[#ftl]
+
+[#--
+    NOTE: Resources WILL require a "Parameters" template pass.
+--]
+
+[#assign mysqlResourceProfiles = {
+    AZURE_DB_MYSQL_SERVER_RESOURCE_TYPE : {
+        "apiVersion" : "2017-12-01",
+        "conditions" : [ "max_length", "name_to_lower", "globally_unique" ],
+        "max_name_length" : 63,
+        "type" : "Microsoft.DBforMySQL/servers",
+        "outputMappings" : {
+            URL_ATTRIBUTE_TYPE : {
+                "Property" : "properties.fullyQualifiedDomainName"
+            }
+        }
+    },
+    AZURE_DB_MYSQL_SERVER_CONFIGURATION_RESOURCE_TYPE : {
+        "apiVersion" : "2017-12-01",
+        "type" : "Microsoft.DBforMySQL/servers/configurations",
+        "outputMappings" : {}
+    },
+    AZURE_DB_MYSQL_SERVER_DATABASE_RESOURCE_TYPE : {
+        "apiVersion" : "2017-12-01",
+        "conditions" : [ "max_length", "name_to_lower", "globally_unique" ],
+        "max_name_length" : 63,
+        "type" : "Microsoft.DBforMySQL/servers/databases",
+        "outputMappings" : {
+            REFERENCE_ATTRIBUTE_TYPE : {
+                "Property" : "id"
+            }
+        }
+    },
+    AZURE_DB_MYSQL_SERVER_FIREWALL_RULE_RESOURCE_TYPE : {
+        "apiVersion" : "2017-12-01",
+        "type" : "Microsoft.DBforMySQL/servers/firewallRules",
+        "outputMappings" : {}
+    },
+    AZURE_DB_MYSQL_SERVER_SECURITY_ALERT_POLICY_RESOURCE_TYPE : {
+        "apiVersion" : "2017-12-01",
+        "type" : "Microsoft.DBforMySQL/servers/securityAlertPolicies",
+        "outputMappings" : {}
+    },
+    AZURE_DB_MYSQL_SERVER_VNET_RULE_RESOURCE_TYPE : {
+        "apiVersion" : "2017-12-01",
+        "type" : "Microsoft.DBforMySQL/servers/virtualNetworkRules",
+        "outputMappings" : {}
+    }
+}]
+
+[#list mysqlResourceProfiles as type,profile]
+    [@addResourceProfile
+        service=AZURE_DB_MYSQL_SERVICE
+        resource=type
+        profile=profile
+    /]
+[/#list]
+
+[#macro createMySqlServer
+    id
+    name
+    location
+    createMode
+    adminName
+    adminSecret
+    keyvaultId
+    skuName=""
+    skuTier=""
+    skuCapacity=""
+    skuSize=""
+    skuFamily=""
+    version=""
+    sslEnforcement="Disabled"
+    backupRetentionDays=""
+    geoRedundandBackup=""
+    storageGB=""
+    storageAutogrow=""
+    sourceServerId=""
+    restorePointInTime=""
+    dependsOn=[]]
+
+    [#local disallowedAdminNames = [
+        "azure_superuser",
+        "admin",
+        "administrator",
+        "root",
+        "guest",
+        "public"
+    ]]
+
+    [#if disallowedAdminNames?seq_contains(adminName?lower_case)]
+        [@precondition
+            function="createMySqlServer"
+            context={ "adminName": adminName, "max_length": 63, "length" : adminName?length }
+            detail="Disallowed database administrator account name, or name too long."
+        /]
+    [/#if]
+
+    [#local adminName = replaceAlphaNumericOnly(adminName)]
```

Comment:
This should probably be another precondition error instead of a magic handling. Otherwise the admin name in your state will be different from the one actually provisioned. 

I.e. if I put in my config `big-admin` it will become `bigadmin` and I won't know in my db connection strings and it will be confusing when looking in my declarative definition 

This instead could be 
```suggestion
    [#if replaceAlphaNumericOnly(adminName) != adminName ]
        [@precondition
            function="createMySqlServer"
            context={ "adminName": adminName }
            detail="admin name can only contain alphanumeric characters"
        /]
    [/#if
```

Reply:
Again plagiarized from postgres. I'll update both